### PR TITLE
Ignore empty `choices` in openai autolog

### LINKED ISSUE #13361: [BUG] IndexError: list index out of range in `_stream_output_logging_hook` in `openai` flvao
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Databricks

### Willingness to contribute

Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.

### MLflow version

MLflow == '2.15.1'

### System information

Does not matter

### Describe the problem

I get `IndexError: list index out of range`, when streaming output on MLflow logged model built in `langchain` flavor with `AzureChatOpenAI` model

I found an [issue](https://github.com/langchain-ai/langchain/issues/6462#issuecomment-1607622266) in Langchain package where they had to solve the same problem.

The `choices` attribute is empty list in first chunk, hence why the following code throws the error.

https://github.com/mlflow/mlflow/blob/b6987a87e45770e0c78f95a8890e2829e748baa2/mlflow/openai/_openai_autolog.py#L153-L156

image

### Tracking information

_No response_

### Code to reproduce issue

Not needed

### Stack trace

```python
IndexError: list index out of range
File <command-3971925096739778>, line 1
----> 1 for c in chain.stream(ROUTER_CONFIG.get("input_example")):
      2     print(c)
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-8acdd35d-ae7d-425f-9d08-5963f156bef4/lib/python3.11/site-packages/langchain_core/runnables/base.py:3262, in RunnableSequence.stream(self, input, config, **kwargs)
   3256 def stream(
   3257     self,
   3258     input: Input,
   3259     config: Optional[RunnableConfig] = None,
   3260     **kwargs: Optional[Any],
   3261 ) -> Iterator[Output]:
-> 3262     yield from self.transform(iter([input]), config, **kwargs)
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-8acdd35d-ae7d-425f-9d08-5963f156bef4/lib/python3.11/site-packages/langchain_core/runnables/base.py:3249, in RunnableSequence.transform(self, input, config, **kwargs)
   3243 def transform(
   3244     self,
   3245     input: Iterator[Input],
   3246     config: Optional[RunnableConfig] = None,
   3247     **kwargs: Optional[Any],
   3248 ) -> Iterator[Output]:
-> 3249     yield from self._transform_stream_with_config(
   3250         input,
   3251         self._transform,
   3252         patch_config(config, run_name=(config or {}).get("run_name") or self.name),
   3253         **kwargs,
   3254     )
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-8acdd35d-ae7d-425f-9d08-5963f156bef4/lib/python3.11/site-packages/langchain_core/runnables/base.py:2056, in Runnable._transform_stream_with_config(self, input, transformer, config, run_type, **kwargs)
   2054 try:
   2055     while True:
-> 2056         chunk: Output = context.run(next, iterator)  # type: ignore
   2057         yield chunk
   2058         if final_output_supported:
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-8acdd35d-ae7d-425f-9d08-5963f156bef4/lib/python3.11/site-packages/langchain_core/runnables/base.py:3212, in RunnableSequence._transform(self, input, run_manager, config, **kwargs)
   3209     else:
   3210         final_pipeline = step.transform(final_pipeline, config)
-> 3212 yield from final_pipeline
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-8acdd35d-ae7d-425f-9d08-5963f156bef4/lib/python3.11/site-packages/langchain_core/output_parsers/transform.py:65, in BaseTransformOutputParser.transform(self, input, config, **kwargs)
     49 def transform(
     50     self,
     51     input: Iterator[Union[str, BaseMessage]],
     52     config: Optional[RunnableConfig] = None,
     53     **kwargs: Any,
     54 ) -> Iterator[T]:
     55     """Transform the input into the output format.
     56 
     57     Args:
   (...)
     63         The transformed output.
     64     """
---> 65     yield from self._transform_stream_with_config(
     66         input, self._transform, config, run_type="parser"
     67     )
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-8acdd35d-ae7d-425f-9d08-5963f156bef4/lib/python3.11/site-packages/langchain_core/runnables/base.py:2020, in Runnable._transform_stream_with_config(self, input, transformer, config, run_type, **kwargs)
   2018 input_for_tracing, input_for_transform = tee(input, 2)
   2019 # Start the input iterator to ensure the input Runnable starts before this one
-> 2020 final_input: Optional[Input] = next(input_for_tracing, None)
   2021 final_input_supported = True
   2022 final_output: Optional[Output] = None
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-8acdd35d-ae7d-425f-9d08-5963f156bef4/lib/python3.11/site-packages/langchain_core/runnables/base.py:1290, in Runnable.transform(self, input, config, **kwargs)
   1287             final = ichunk
   1289 if got_first_val:
-> 1290     yield from self.stream(final, config, **kwargs)
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-8acdd35d-ae7d-425f-9d08-5963f156bef4/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py:411, in BaseChatModel.stream(self, input, config, stop, **kwargs)
    404 except BaseException as e:
    405     run_manager.on_llm_error(
    406         e,
    407         response=LLMResult(
    408             generations=[[generation]] if generation else []
    409         ),
    410     )
--> 411     raise e
    412 else:
    413     run_manager.on_llm_end(LLMResult(generations=[[generation]]))
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-8acdd35d-ae7d-425f-9d08-5963f156bef4/lib/python3.11/site-packages/langchain_core/language_models/chat_models.py:391, in BaseChatModel.stream(self, input, config, stop, **kwargs)
    388     self.rate_limiter.acquire(blocking=True)
    390 try:
--> 391     for chunk in self._stream(messages, stop=stop, **kwargs):
    392         if chunk.message.id is None:
    393             chunk.message.id = f"run-{run_manager.run_id}"
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-8acdd35d-ae7d-425f-9d08-5963f156bef4/lib/python3.11/site-packages/langchain_openai/chat_models/base.py:640, in BaseChatOpenAI._stream(self, messages, stop, run_manager, **kwargs)
    638 with response:
    639     is_first_chunk = True
--> 640     for chunk in response:
    641         if not isinstance(chunk, dict):
    642             chunk = chunk.model_dump()
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-8acdd35d-ae7d-425f-9d08-5963f156bef4/lib/python3.11/site-packages/openai/_streaming.py:46, in Stream.__iter__(self)
     45 def __iter__(self) -> Iterator[_T]:
---> 46     for item in self._iterator:
     47         yield item
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-8acdd35d-ae7d-425f-9d08-5963f156bef4/lib/python3.11/site-packages/mlflow/openai/_openai_autolog.py:156, in patched_call.<locals>._stream_output_logging_hook(stream)
    154     output.append(chunk.choices[0].text or "")
    155 elif isinstance(chunk, ChatCompletionChunk):
--> 156     output.append(chunk.choices[0].delta.content or "")
    157 chunks.append(chunk)
    158 yield chunk
```

### Other info / logs

_No response_

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [X] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by harupy**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/harupy/mlflow/pull/13372?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/13372/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 13372
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Fix #13361. The first chunk in a completions/chat-completions stream can be empty. This PR updates openai autolog to handle that.

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: tests/openai/mock_openai.py [hunk 1]
```diff
@@ -92,8 +111,15 @@ async def chat(payload: ChatPayload):
         )
     if payload.stream:
         # SSE stream
+        if EMPTY_CHOICES == payload.messages[0].content:
+            content = (
+                f"data: {json.dumps(d)}\n\n" async for d in chat_response_stream_empty_choices()
+            )
+        else:
+            content = ((f"data: {json.dumps(d)}\n\n" async for d in chat_response_stream()),)
```
Comment (harupy) [inline tests/openai/mock_openai.py:119]:
```suggestion
            content = (f"data: {json.dumps(d)}\n\n" async for d in chat_response_stream())
```

**Review by BenWilson2 — APPROVED** [review]
Seems reasonable and thoroughly tested. 
Question (non blocking - just will be good to verify if possible) - is there any case where important metadata is returned in an empty choices chunk that is otherwise not present in a chunk that has content? 
I doubt this is the case, but it could be good to verify (if this is even reliably repeatable, which is may not be).

**Comment by harupy** [comment]
@BenWilson2 Good question. Not sure if this is the case for OpenAI, but AzureOpen's empty chunk has `prompt_annotations`, which seems important:

https://github.com/mlflow/mlflow/issues/13361

image

**Comment by BenWilson2** [comment]
oooof is it possible to extract that for tracing purposes?

**Comment by BenWilson2** [comment]
FWIW, this is a bug fix PR and I think that it's valuable to merge and fix. I think we can address tracing enhancements for this odd behavior as a follow up.

**Comment by harupy** [comment]
yea it's a separate issue. We can merge this PR.

**Pull request merged**